cmd/tegola/cmd/cache.go

```
@@ -279,35 +167,234 @@ var cacheCmd = &cobra.Command{
+	//	track how long the tile generation is taking
+	t := time.Now()
+
+	//	lookup the Map
+	m, err := atlas.GetMap(mt.MapName)
+	if err != nil {
+		log.Fatalf("error seeding tile (%+v): %v", mt.Tile, err)
+	}
+
+	z, x, y := mt.Tile.ZXY()
+
+	//	filter down the layers we need for this zoom
+	m = m.FilterLayersByZoom(x)
+
+	//	check if overwriting the cache is not ok
+	if !cacheOverwrite {
+		//	lookup our cache
+		c := atlas.GetCache()
+		if c == nil {
+			log.Fatalf("error fetching cache: %v", err)
+		}
+
+		//	cache key
+		key := cache.Key{
+			MapName: mt.MapName,
+			Z:       z,
+			X:       x,
+			Y:       y,
+		}
+
+		//	read the tile from the cache
+		_, hit, err := c.Get(&key)
+		if err != nil {
+			log.Fatal("error reading from cache: %v", err)
+		}
+		//	if we have a cache hit, then skip processing this tile
+		if hit {
+			log.Infof("cache seed set to not overwrite existing tiles. skipping map (%v) tile (%v/%v/%v)", mt.MapName, z, x, y)
+			return
+		}
+	}
+
+	//	set tile buffer if it was configured by the user
+	if conf.TileBuffer > 0 {
+		mt.Tile.Buffer = float64(conf.TileBuffer)
+	}
+
+	//	seed the tile
+	if err = atlas.SeedMapTile(ctx, m, z, x, y); err != nil {
+		log.Errorf("error seeding tile (%+v): %v", mt.Tile, err)
+		return
+	}
+
+	//	TODO: this is a hack to get around large arrays not being garbage collected
+	//	https://github.com/golang/go/issues/14045 - should be addressed in Go 1.11
+	runtime.GC()
+
+	log.Infof("seeding map (%v) tile (%v/%v/%v) took: %v", mt.MapName, z, x, y, time.Now().Sub(t))
+}
+
+func purgeWorker(mt MapTile) {
+
+	z, x, y := mt.Tile.ZXY()
+
+	log.Infof("purging map (%v) tile (%v/%v/%v)", mt.MapName, z, x, y)
+
+	//	lookup the Map
+	m, err := atlas.GetMap(mt.MapName)
+	if err != nil {
+		log.Fatalf("error seeding tile (%+v): %v", mt.Tile, err)
+	}
+
+	//	purge the tile
+	ttile := tegola.NewTile(mt.Tile.ZXY())
+	if err = atlas.PurgeMapTile(m, ttile); err != nil {
+		log.Errorf("error purging tile (%+v): %v", mt.Tile, err)
```

return error instead of logging info